Add experimental plumbing for hardware accelerated rendering operations with the software renderer

tronical opened the pull request:
cc #5774

tronical commented:
cc @scristall-bennu @0x6e

tronical commented on api/cpp/include/slint-platform.h:
```diff
@@ -680,6 +731,114 @@ class SoftwareRenderer : public AbstractRenderer
+                    [](void *self, int16_t x, int16_t y, int16_t width, int16_t height,
+                       int16_t span_y, const cbindgen_private::CppInternalTexture *internal_texture,
+                       uint32_t colorize, uint8_t alpha, int32_t screen_rotation_degrees) {
+                        auto *buffer = reinterpret_cast<TargetPixelBuffer<Rgb8Pixel> *>(self);
```
Ohh, good one. Will push a fix. Thanks!

tronical commented on api/cpp/include/slint-platform.h:
```diff
@@ -538,6 +538,68 @@ struct Rgb565Pixel
+    uint16_t source_offset_x;
+    /// The offset within the texture to start reading pixels from in the y direction. The
+    /// offset is specified in 8:8 fixed point format.
+    uint16_t source_offset_y;
```
Yes indeed. Thanks!

ogoffart commented on internal/core/software_renderer.rs:
```diff
@@ -1128,21 +1240,85 @@ impl<T: TargetPixel> RenderToBuffer<'_, T> {
+    ) {
+        let mut line = geometry.min_y();
+        while let Some(mut next) =
+            region_line_ranges(&self.dirty_region, line, &mut self.dirty_range_cache)
```
region_line_ranges tries to split the ranges so that they span on lines. for the line by line rendering.
But here, i think it would be wiser to simply iterate over the rectangle in the region. I think they shouldn't overlap.

Example:
Imagine the following dirty region
```
+----+    +----+
|  A |    |  B |
|    |    +----+
|    |    
+----+
```

This will then generate a region for the upper part of A, B, and the lower part of A.
But it'd be better to "simply" draw A, then B.  Not?

tronical replied:
Sounds good to me.

tronical replied:
I hope I understood this correctly and applied this to rectangle and texture fills.

ogoffart replied:
I don't think you understood.
My point was that, in this funciton, instead of calling `region_line_ranges`,  just call self.dirty_region.iter_box(),  but the rectangles can overlap so that is not a good idea.

tronical replied:
I see. What do you suggest instead?

tronical commented on internal/core/software_renderer.rs:
```diff
@@ -82,7 +80,7 @@ impl RenderingRotation {
     }
     /// Angle of the rotation in degrees
-    fn angle(self) -> f32 {
+    pub fn angle(self) -> f32 {
```
Ah yes. It can't be `pub(crate)` because it's accessed from the cpp crate, but I can work around it and duplicate this function, alternatively. I think the signature and docs are fine as-is, but I can do the workaround if you prefer.

ogoffart replied:
I think it's fine, i just want to make sure we make a conscious choice to make this public.

tronical commented on internal/core/software_renderer.rs:
```diff
@@ -365,6 +363,110 @@ fn region_line_ranges(
     next_validity
 }
 
+mod private_api {
```
Agreed :)

tronical commented on internal/core/software_renderer.rs:
```diff
@@ -541,26 +666,14 @@ impl SoftwareRenderer {
-
+                PremultipliedRgbaColor::blend(&mut bg, background.color().into());
                 renderer.actual_renderer.processor.dirty_region = dirty_region.clone();
+                renderer.actual_renderer.processor.process_rectangle(
```
Ah yes, indeed. I'll add a composition mode to handle this.

tronical replied:
I've tried to implement this now, but there's a caveat/question, see process_rectangle_impl.

tronical commented on internal/core/software_renderer.rs:
```diff
@@ -1128,21 +1163,117 @@ impl<T: TargetPixel> RenderToBuffer<'_, T> {
+                    match composition_mode {
+                        CompositionMode::Source => buffer.line_slice(l as usize)
+                            [begin as usize..end as usize]
+                            .fill(<T as TargetPixel>::from_rgb(color.red, color.green, color.blue)),
```
This doesn't seem entirely correct to me, as it doesn't copy over the alpha and it doesn't divide the pre-multipled alpha.

Should I add a `fill_slice()` to `TargetPixel`, similar to `blend_slice`?

ogoffart replied:
Right, this is not correct as this loses the alpha.

You can do what was done before:
```rs
let mut bg = T::background();
T::blend(&mut bg, color);

buffer.line_slice(/*...*/)[/*...*/].fill(bg)
```

But adding a fill_slice would also work, i guess.

ogoffart commented on internal/core/software_renderer.rs:
```diff
@@ -1082,14 +1118,13 @@ trait ProcessScene {
 }
 
-impl<T: TargetPixel> RenderToBuffer<'_, T> {
+impl<T: TargetPixel, B: target_pixel_buffer::TargetPixelBuffer<Pixel = T>> RenderToBuffer<'_, B> {
```
Do we need T?
Can it be done with

```suggestion
impl<B: target_pixel_buffer::TargetPixelBuffer> RenderToBuffer<'_, B> {
```

And then use `P::Pixel` instead of T

tronical replied:
Excellent idea. Done :)

ogoffart commented on internal/core/software_renderer.rs:
```diff
@@ -541,26 +588,15 @@ impl SoftwareRenderer {
                 drop(i);
 
-                let mut bg = TargetPixel::background();
+                let mut bg = PremultipliedRgbaColor::background();
```
```suggestion
```

ogoffart commented on internal/core/software_renderer.rs:
```diff
@@ -541,26 +588,15 @@ impl SoftwareRenderer {
-                    line = next;
-                }
-
+                PremultipliedRgbaColor::blend(&mut bg, background.color().into());
```
```suggestion
```

ogoffart commented on internal/core/software_renderer.rs:
```diff
@@ -541,26 +588,15 @@ impl SoftwareRenderer {
                 renderer.actual_renderer.processor.dirty_region = dirty_region.clone();
+                renderer.actual_renderer.processor.process_rectangle_impl(
+                    PhysicalRect::from_size(euclid::Size2D::new(size.width, size.height)),
+                    bg,
```
```suggestion
                    background.color().into(),
```

ogoffart commented on internal/core/software_renderer.rs:
```diff
@@ -1128,21 +1163,117 @@ impl<T: TargetPixel> RenderToBuffer<'_, T> {
-                extra_right_clip,
-            );
-        });
+        if !self.buffer.draw_texture(
```
That was better before. If you are not passing the dirty region to the function, how is it going to clip for the dirty region?
So I see two possibility:
 1. Put pack the call to draw_texture within foreach_region
 2. Or pass the dirty region to the draw_texture function and let the implementer do the clipping.

I'd think 1. is the most convenient, unless there is a way to do hardware clip, but i don't think that's the case on most hardware i know.

tronical replied:
👍 I reverted back to 1.

ogoffart commented on internal/core/software_renderer.rs:
```diff
@@ -1082,14 +1118,13 @@ trait ProcessScene {
 
-impl<T: TargetPixel> RenderToBuffer<'_, T> {
+impl<T: TargetPixel, B: target_pixel_buffer::TargetPixelBuffer<Pixel = T>> RenderToBuffer<'_, B> {
     fn foreach_ranges(
```
Maybe this could be implemented in terms of foreach_region to avoid code duplication?

tronical replied:
Good idea!

ogoffart commented on internal/core/software_renderer.rs:
```diff
@@ -1128,21 +1163,117 @@ impl<T: TargetPixel> RenderToBuffer<'_, T> {
+            texture.extra.rotation,
+            CompositionMode::default(),
+        ) {
+            self.foreach_region(&geometry, |buffer, rect, extra_left_clip, extra_right_clip| {
+                let begin = rect.min_x();
+                let end = rect.max_x();
+                for l in rect.min_y()..rect.max_y() {
```
that's just foreach_ranges

ogoffart commented on internal/core/software_renderer/target_pixel_buffer.rs:
```diff
@@ -0,0 +1,81 @@
+/// to delegate rendering further to hardware-provided 2D acceleration units, such as DMA2D or PXP.
+pub trait TargetPixelBuffer {
+    /// The pixel type the buffer represents.
+    type Pixel: TargetPixel;
```
elsewhere we just have

```suggestion
    type TargetPixel: TargetPixel;
```
 (like in LineBufferProvider)

ogoffart commented on internal/core/software_renderer/target_pixel_buffer.rs:
```diff
@@ -0,0 +1,81 @@
+        _color: PremultipliedRgbaColor,
+        _composition_mode: CompositionMode,
+    ) -> bool {
+        false
```
instead of returning bool, we could make the default implementation call the fallback.
But then the implementer has no way to call the fallback only in some cases. so i guess it's good the way you did it.

tronical replied:
That was my thinking, too.

tronical commented:
I hope I addressed all feedback. There's the open question about the `angle()` function: I'm happy to go either way, make it public or not. Let me know what you prefer.

ogoffart commented on internal/core/software_renderer.rs:
```diff
@@ -454,6 +478,34 @@ impl SoftwareRenderer {
+        &self,
+        buffer: &mut impl target_pixel_buffer::TargetPixelBuffer,
+    ) -> PhysicalRegion {
+        let pixel_stride = buffer.line_slice(0).len();
```
That's not technically a stride.

Since this is only used to assert here that the buffer is big enough, I think it'd be fine to move that assert back to the render() function.
I'm not sure the case where the window_item is None is actually not dead code anyway. But we probably can get the size from the window's size

tronical replied:
Good point!

tronical replied:
One second thought, I see value in having this in this function instead of just `render()`, as even a provided `TargetPixelBuffer` could have wrong dimensions. But I'll rename the variable, that's indeed not the stride.

ogoffart commented on internal/core/software_renderer.rs:
```diff
@@ -1082,18 +1116,35 @@ trait ProcessScene {
+        mut f: impl FnMut(i16, &mut [B::TargetPixel], i16, i16),
+    ) {
+        self.foreach_region(geometry, |buffer, rect, extra_left_clip, extra_right_clip| {
+            for l in rect.min_y()..rect.max_y() {
```
I think this can be written
```suggestion
            for l in rect.y_range() {
```

tronical replied:
Nice - TIL :)

ogoffart commented on internal/core/software_renderer.rs:
```diff
@@ -1129,20 +1178,84 @@ impl<T: TargetPixel> RenderToBuffer<'_, T> {
+                for l in rect.min_y()..rect.max_y() {
+                    match composition_mode {
+                        CompositionMode::Source => {
+                            let mut fill_col = B::TargetPixel::background();
+                            B::TargetPixel::blend(&mut fill_col, color);
```
It'd be faster to put that outside of the loop.  (meaning also swapping the `match` and the `for`

tronical replied:
I was contemplating that and thinking the compiler will do that automatically. But swapping match and for is a good idea anyway.

ogoffart reviewed (approved):
Would be nice if you could also check if there is any difference in the printerdemo_mcu benchmark.

This patch adds a bit of overhead, but hopefully it is negligible.
